### ISSUE TITLE
Error when 2 zaken are created at exactly the same moment

### ISSUE THREAD

**Issue by petervanmanen** [open-zaak/open-zaak#1271]
**General information**
Open Zaak version: 1.7.3-maykinlatest (with 3 openzaak nodes)

**Describe the bug**
When loadtesting we ran into an issue where if two threads create a new zaak (without specifying an identificatie) at exactly the same moment an error occurs:  HTTPError('500 Server Error: Internal Server Error for url: https://openzaak-zgw.test.denhaag.nl/zaken/api/v1/zaken')

It appears that two processes can be awarded the same identificatie and the first one to use it is succesfull, the second one gets an error the the identificatie is not unique. However openzaak itself generates the identificatie.

**To Reproduce**
1. Have script that creates a lot of zaken in multiple threads
2. Wait until the collision occurs

**Expected behavior**
Openzaak is threadsafe and multiple threads can create zaken at the same time.

**Comment by joeribekker** [comment]
Judging by the screenshot, you provided the zaak "identificatie" yourself. Open Zaak uses UUID4 to create unique zaaknumbers (with near-zero chance of colissions). Are you sure this collision was caused by Open Zaak (and some custom number generator)?

**Comment by petervanmanen** [comment]
@joeribekker this is the post request I use
{"bronorganisatie": "111222333","omschrijving": "BLAABLAA","toelichting": "BLAABLAA","zaaktype": "https://openzaak-zgw.blabla.nl/catalogi/api/v1/zaaktypen/b61bce86-f2ac-4aff-a155-d4d51884395d","verantwoordelijkeOrganisatie": "111222333","startdatum": "2022-09-29"}

So no identificatie provided

**Comment by joeribekker** [comment]
Hmm, you're right. My memory is incorrect.

Never had this before, even when importing/migrating large amounts of cases. Maybe those were always done in serial.

**Comment by sergei-maertens** [comment]
because of transaction isolation, this is indeed a risk - the only way we can prevent this is probably by taking a lock on the entire database table, and I think this will bring bigger (performance) problems with it.

I think it'd make more sense to translate this particular unique constraint violation as a HTTP 409 Conflict, so the client can retry.

Another way would be to use a different number generation algorithm (non-sequential/randomized) with lower chances of collisions, or automatically retry a number of times if we run into conflict.

**Comment by petervanmanen** [comment]
@sergei-maertens today I ran a test, it ran for 20 minutes. In these 20 minutes it tried creating 280 zaken, it failed 86 times. 
![image](https://user-images.githubusercontent.com/10403912/193813810-03c477ea-590c-4b3b-9e65-6a48bf4ef6b1.png)

It appears to me that the transaction in openzaak takes way to long, it's not just an error that occurs in very rare occasions but even with moderate load it happens quite often...

**Comment by sergei-maertens** [comment]
5.3s for a ZAAK create operation (median) immediately stands out to me. First action when picking up this ticket will be to perform similar profiling to see what is happening on this endpoint, because that really shouldn't take that long. That should in turn indeed decrease the probability of these conflicts happening because the DB transaction should not be taking that long.

I haven't thought in a particular solution direction yet, I have a couple of ideas, but my previous comment was mostly a braindump so I don't forget when we eventually get to picking this up :-)

One extra thing I've thought of, is before committing the transaction checking if the case number is still available and/or handling the unique constraint violation and retrying with a new number (up to X number of times, for example).

But first, I think the root cause for those long transaction durations should be found.

**Comment by petervanmanen** [comment]
@sergei-maertens with regards to: _I checked APM and I see a lot of DB queries, there's definitely room for improvement there :)_

Do you mean there is a lot of room for us to improve the database performance or that there is a lot of room to improve the queries and logic in openzaak?

**Comment by sergei-maertens** [comment]
There is room in Open Zaak to investigate why this many queries are being done and try to optimize that :)

**Comment by sergei-maertens** [comment]
While the queries are being optimized, I also experimented a bit with pushing the identification generation down to the database level - this query seems to achieve the desired result:

```sql
select 
  concat(
    'ZAAK-2022-', 
    to_char(
      max(
        split_part(identificatie, '-', 3)::integer
      ) + 1, 
      'FM0000000000'
    )
  ) as next 
FROM zaken_zaak 
WHERE identificatie ~ 'ZAAK-2022-[0-9]{10}';
```

This should easily be wrapped in a Django expression for these functions so we can let the database generate the next identification instead of doing this in python/django which is much more susceptible to these race conditions.

**Comment by sergei-maertens** [comment]
definitely plenty of transactions, and due to the sheer amount of queries transactions still take too long, but it further reduces the probability of race conditions, as other transactions may complete before the current one trying to insert but that will not cause problems if you insert with an expression that is only evaluated at insertion time

**Comment by sergei-maertens** [comment]
https://forum.djangoproject.com/t/insert-into-select-query-support-in-django/14648/3 meh, django doesn't support `insert into select` :(

**Comment by sergei-maertens** [comment]
Discussed with Alex - this is too much work (even a "hotfix") for a patch release. It's in a state that we'll be able to include it in 1.8.0 though.

**Comment by sergei-maertens** [comment]
#1299 improves performance, while the actual race condition is fixed via #1300

Some quick and dirty load testing for concurrent zaak creations (only) with UWSGI 4 processes 4 threads resulted in:

* Open Zaak 1.7.5: ~75% failure rate, 20-30rps
* #1299 branch: 0% failure rate and 40-45 rps

So those results look very promising

**Comment by alextreme** [comment]
Oooo very nice, double the performance and a successrate that feels a lot more solid :+1:
